marklogic-data-hub-spark-connector/src/test/java/com/marklogic/hub/spark/sql/sources/v2/reader/ReadDataWithCustomEndpointsTest.java

```diff
@@ -0,0 +1,78 @@
+package com.marklogic.hub.spark.sql.sources.v2.reader;
+
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import com.marklogic.hub.spark.sql.sources.v2.Options;
+import org.apache.spark.sql.catalyst.InternalRow;
+import org.apache.spark.sql.types.StructType;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+import java.util.List;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+public class ReadDataWithCustomEndpointsTest extends AbstractSparkReadTest {
+
+    @BeforeEach
+    void setUp() {
+        installCustomEndpoint("custom-read-endpoints/initializeRead.api", "custom-read-endpoints/initializeRead.sjs");
+        installCustomEndpoint("custom-read-endpoints/customReadRows.api", "custom-read-endpoints/customReadRows.sjs");
+    }
+
+    @Test
+    public void testCustomSchema() {
+        setupCustomers();
+        Options options = newOptions().withView("Customer")
+            .withInitializeReadApiPath("/custom-read-endpoints/initializeRead.api");
+        HubDataSourceReader dataSourceReader = new HubDataSourceReader(options.toDataSourceOptions());
+        StructType testStructType = dataSourceReader.readSchema();
+
+        assertTrue(testStructType.toList().length() == 2, "Custom schema not as expected.");
+        for(int i=0; i<testStructType.toList().length(); i++) {
+            String name = testStructType.toList().apply(0).name();
+            assertTrue(name.equals("myName") || name.equals("myId"));
+        }
+    }
+
+    @Test
+    public void testHasEndpointStateWithoutApiPath() {
+        setupCustomers();
+        ObjectNode readrowsendpointparams = objectMapper.createObjectNode();
+        readrowsendpointparams.put("endpointState", "testValue");
+        Options options = newOptions().withView("Customer").withReadRowsEndpointparams(readrowsendpointparams);
+        IllegalArgumentException ex = assertThrows(IllegalArgumentException.class,
+            () -> new HubDataSourceReader(options.toDataSourceOptions()));
+        assertEquals("Cannot set endpointState in readrowsendpointparams option unless apiPath is defined as well.",
+            ex.getMessage());
+    }
+
+    @Test
+    public void testHasEndpointConstants() {
+        setupCustomers();
+        ObjectNode readrowsendpointparams = objectMapper.createObjectNode();
+        readrowsendpointparams.put("endpointConstants", "testValue");
+        Options options = newOptions().withView("Customer").withReadRowsEndpointparams(readrowsendpointparams);
+        IllegalArgumentException ex = assertThrows(IllegalArgumentException.class,
+            () -> new HubDataSourceReader(options.toDataSourceOptions()));
+        assertEquals("Cannot set endpointConstants in readrowsendpointparams option; can only set apiPath and endpointState.",
+            ex.getMessage());
+    }
+
+    @Test
+    public void testReadRowsEndpointparams() {
```

Comment:
This test isn't verifying that the customReadRows.api endpoint is being used, because it's asserting that rows is empty. Rows is empty because no customers have been loaded. 

The test class I added - ReadWithCustomEndpointsTest - is failing because HubInputPartitionReader needs to look at the options to determine if it should use a custom API path or not. That test should suffice too for the happy path (you'll still want test methods for when endpointState is provided by not apiPath, and also for when endpointConstants is erroneously provided; I recommend moving both of those into ReadWithCustomEndpointsTest).